documentation/projects/proposals/ingestion_server_removal/20240319-project_proposal.md

```diff
@@ -0,0 +1,105 @@
+
+Currently, our data refresh process is _orchestrated_ by Airflow, but the work
+is actually performed on a bespoke ingestion server. This server is relatively
+complex compared to our needs, difficult to maintain, and makes the data refresh
+process unnecessarily opaque. Because the refresh is triggered in a single task
+on the ingestion server, it is not possible to track the progress in Airflow or
+retry it from a point of failure. Progress must be followed through periodic
+Slack notifications and inspection of logs in Cloudwatch.
+
+In this project, we will move all operational pieces of the data refresh into
+Airflow and completely remove the ingestion server.
+
+## Goals
+
+<!-- Which yearly goal does this project advance? -->
+
+Moving the process entirely into Airflow will have the following advantages for
+the critical data refresh process:
+
+- Easier iteration, as changes to the implementation can be deployed immediately
+  by simply merging code in the catalog
+- Increased reusability of code, particularly among the growing number of
+  Elasticsearch DAGs
+- Increased observability of the process: we'll have the ability to see exactly
+  what step the refresh is on by inspecting the running DAG, with much more
+  granularity even than the existing Slack notifications. This also includes
+  much more convenient access to timing information
+- Ability to pause the process before the next step starts, or retry a failed
+  refresh from the point of failure
+- Easier to introduce a staging data refresh DAG, which currently does not exist
+  but which would be invaluable for testing of the process
+
+Removing the ingestion server also has the following advantages from an
+infrastructure perspective:
+
+- Fewer services to manage overall; no maintenance of a bespoke server
+- Reduced costs (no requirement for 24/7 EC2 instance)
+- Improved deployments
+
+## Requirements
+
+<!-- Detailed descriptions of the features required for the project. Include user stories if you feel they'd be helpful, but focus on describing a specification for how the feature would work with an eye towards edge cases. -->
+
+- The data refresh process should be managed entirely by Airflow
+  - The expensive reindexing step currently makes use of a fleet of
+    `indexer workers`, which will still be required and which can not be
+    provided by Airflow directly. These should be managed by Airflow using ECS
+    or EC2 operators, but as much of the reindexing logic as possible should
+    remain in the catalog itself. More specifics are left to the implementation
+    plan.
+- There should be no increase in the time it takes to perform a data refresh
+
+## Success
+
+<!-- How do we measure the success of the project? How do we know our ideas worked? -->
+
+The project will be considered a success when at least one full audio and image
+data refresh are run successfully with the new implementation, and the ingestion
+server is removed.
```

Comment:
Would it be okay to be more specific here with the ingestion server pieces of the success criteria? Specifically that we should be able to deprovision the ingestion server infrasturcture resources and delete all associated code in the main & infrastructure repositories. (if you think that's reasonable)